File: internal/confidence/confidence.go
package confidence

import (
	"math"
	"strings"
	"swarmvault-native/internal/types"
)

func NodeConfidence(sourceCount int) float64 {
	return math.Min(0.5+float64(sourceCount)*0.15, 0.95)
}

func EdgeConfidence(claims []types.SourceClaim, conceptName string) float64 {
	lower := strings.ToLower(conceptName)
	var relevant []types.SourceClaim
	for _, c := range claims {
		if strings.Contains(strings.ToLower(c.Text), lower) {
			relevant = append(relevant, c)
		}
	}
	if len(relevant) == 0 {
		return 0.5
	}
	sum := 0.0
	for _, c := range relevant {
		sum += c.Confidence
	}
	return sum / float64(len(relevant))
}

func ConflictConfidence(claimA types.SourceClaim, claimB types.SourceClaim) float64 {
	return math.Min(claimA.Confidence, claimB.Confidence)
}

File: packages/engine/src/confidence.ts
import { runGoSidecarSync } from "./subprocess.js";
import type { SourceClaim } from "./types.js";

export function nodeConfidence(sourceCount: number): number {
  if (process.env.USE_GO_PORT === "true") {
    return runGoSidecarSync<number>("confidence", { action: "nodeConfidence", args: { sourceCount } });
  }

  return Math.min(0.5 + sourceCount * 0.15, 0.95);
}

export function edgeConfidence(claims: SourceClaim[], conceptName: string): number {
  if (process.env.USE_GO_PORT === "true") {
    return runGoSidecarSync<number>("confidence", { action: "edgeConfidence", args: { claims, conceptName } });
  }

  const lower = conceptName.toLowerCase();
  const relevant = claims.filter((c) => c.text.toLowerCase().includes(lower));
  if (!relevant.length) {
    return 0.5;
  }
  return relevant.reduce((sum, c) => sum + c.confidence, 0) / relevant.length;
}

export function conflictConfidence(claimA: SourceClaim, claimB: SourceClaim): number {
  if (process.env.USE_GO_PORT === "true") {
    return runGoSidecarSync<number>("confidence", { action: "conflictConfidence", args: { claimA, claimB } });
  }

  return Math.min(claimA.confidence, claimB.confidence);
}
